### ISSUE TITLE
[FIX-015] Make DoRA's zero-row norm guard effective in FP16

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#225]
**Category:** PEFT integrity · **Priority:** P2 · **Size:** S

**Evidence status:** Reproduced numerical defect.

## Outcome and scope

Make DoRA row normalization finite for zero/tiny FP16 rows and differentiable for learned nonzero updates. Carry normalization in at least FP32 while preserving FP64 and the effective weight/output dtype. Keep the magnitude/LoRA parameterization and dropout semantics; this is independent of allocation placement because the reproduction converts the entire wrapper.

## Current behavior / evidence

**Context and reproduction:** Zero the weight of `nn.Linear(4,4,bias=False)`, construct `DoRALinear(base,r=2)`, and convert the entire wrapper to half (so [FIX-003](https://github.com/thekaveh/NNx/issues/202) is not involved). A forward on all-one half inputs returns all NaNs although `base(x)` returns finite zeros. FP16 rounds the purported guard `1e-8` to zero, so the normalization still divides zero by zero. Ordinary nonzero half DoRA weights pass the control probe. This arises after pruning, explicit zero initialization, or a zero row in a loaded base.

## Repository map

- [`src/nnx/peft/dora.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/peft/dora.py#L90) — `norm.clamp_min(1e-8)` currently executes in V’s half dtype.
- [`src/nnx/peft/dora.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/peft/dora.py#L87) — Magnitude initialization establishes identity-at-zero-adapter behavior.
- [`src/nnx/peft/lora.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/peft/lora.py#L103) — B zero initialization is essential to the base-parity oracle.
- [`tests/test_peft_dora.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_peft_dora.py#L48) — Ordinary FP32 initial parity and zero-bias coverage.
- [`tests/test_peft_dora.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_peft_dora.py#L77) — Keep only adapter matrices/magnitude trainable.

## Implementation approach

1. Perform normalization on a differentiable promoted representation of the effective weight V: FP32 for FP16/BF16 and unchanged FP64 for doubles. Use a positive clamp representable in the accumulation dtype. Do not merely increase a half-precision epsilon enough to hide zero division, which can distort small valid rows.
2. Compute the row norm, clamp, division and magnitude rescaling coherently before casting the effective weight back for `F.linear`. Maintain the caller’s dtype/autocast contract and bias compatibility. Preserve zero rows as zero at initialization: replacing a zero denominator must not invent a nonzero direction.
3. Keep gradients through A, B and magnitude; do not detach normalized weights or rewrite Parameters during forward. Retain DoRA’s existing DropConnect-style dropout on the composed update rather than changing it into LoRA input dropout. Ensure no unconditional FP32 cast truncates explicit FP64.
4. Compare initial and learned-state behavior separately. Zero magnitude can legitimately give zero gradients at a zero row; require finite gradients and a correct reference, not an arbitrary nonzero-gradient assertion. Nonzero learned updates must demonstrate genuine differentiability.

## Acceptance criteria

- [ ] A zero row in an FP16 base gives finite output equal to the base at adapter initialization.
- [ ] Mixed zero/nonzero rows and sufficiently small representable rows produce finite forwards and gradients.
- [ ] FP32/BF16 and FP64 parity remains correct under their tolerances.
- [ ] Nonzero learned adapter/magnitude updates remain differentiable and finite.

## Validation and completion

Minimal existing-API fixture isolates normalization from FIX-003:

```python
import torch
from torch import nn
from nnx import DoRALinear

base = nn.Linear(4, 4, bias=False)
with torch.no_grad():
    base.weight.zero_()
adapted = DoRALinear(base, r=2).half()
x = torch.ones(2, 4, dtype=torch.float16, requires_grad=True)
out = adapted(x)
assert out.dtype == torch.float16 and torch.isfinite(out).all()
torch.testing.assert_close(out, base(x), atol=0, rtol=0)
out.float().sum().backward()
assert x.grad is not None and torch.isfinite(x.grad).all()
assert all(p.grad is None or torch.isfinite(p.grad).all()
           for p in adapted.parameters())
```

Add `test_dora_half_zero_rows_are_finite`, `test_dora_small_rows_match_promoted_reference`, and `test_dora_nonzero_learned_update_gradients`. Cover all-zero, one-zero-among-nonzero and small representable rows, with/without bias. Reference `V` normalization and rescaling in FP32/FP64 with the same epsilon and final cast; compare output and input/A/B/magnitude gradients at well-conditioned nonzero points. Include FP32/BF16/FP64 and real supported CPU/CUDA cases. Record backend skips and do not promise arbitrary extreme-magnitude overflow resistance.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_peft_dora.py tests/test_peft_lora.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Composition after normalization.** The repaired [DoRALinear.forward](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/peft/dora.py) returns an effective weight consumed by `functional.linear`, often followed by another projection. Promote only the normalization arithmetic required for stability, then preserve the caller-facing dtype/device so a repaired half layer does not break the next half layer. Exercise `apply_dora_to` in a small classifier and a real gradient/update path, not only `isfinite` on the wrapper output. Whole-wrapper conversion after wrapping can isolate this defect from independent constructor-placement [FIX-003](https://github.com/thekaveh/NNx/issues/202).

- **Training and persistence.** The learned `magnitude` participates in the zero-row contract alongside A/B and base bias. A zero row may legitimately have zero gradients at initialization; require finite gradients and differentiability of nonzero learned directions, not artificial nonzero gradients for every zero tensor. Save and reload the complete wrapped module state, including magnitude, then repeat finite eval and a supported optimizer update. [LoRA-only save/load helpers](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/peft/lora.py) still omit magnitude, so they are not a complete DoRA round-trip oracle. Keep key names, tensor shapes and source-base freezing unchanged; no new epsilon setting or serialized precision policy is needed.

- **Executable dtype example.** Extend [Example 07](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/07_lora_finetuning.py) with a bounded DoRA helper containing mixed zero/nonzero base rows, a following projection and one tiny optimizer update. Use a float32 reduction of outputs for a CPU-half regression loss when needed instead of relying on unsupported half MSE backward; run actual supported FP16/BF16 backends separately. Check the complete-state round trip and unchanged frozen base. Document the zero-row/tiny-row behavior and full-state persistence requirement without changing DoRA’s matrix-dropout semantics or implying that all optional mixed-precision kernels are supported.

### Samples and executable coverage

- [`examples/07_lora_finetuning.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/07_lora_finetuning.py): Add a small mixed-zero-row DoRA composition helper with a following same-dtype projection, one optimizer update using a supported loss reduction, frozen-base checks and full state_dict reload. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k dora_zero_row_composition. CPU FP32 and supported CPU FP16 path using float32 loss reduction; optional real BF16/CUDA lanes separately. Convert the full wrapper after wrapping to isolate [FIX-003](https://github.com/thekaveh/NNx/issues/202) and avoid full main. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Explain finite zero/tiny-row DoRA normalization, output dtype preservation and meaningful gradient expectations; reiterate that magnitude needs full module state persistence. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] A public apply_dora_to classifier followed by another same-dtype projection completes finite forward/backward/update with mixed zero/nonzero rows, preserving external dtype and frozen base values.
- [ ] A full wrapped-module state_dict round trip including learned magnitude reproduces finite eval outputs and permits a subsequent supported update without renaming adapter tensors.
- [ ] The integration oracle distinguishes finite legitimate zero gradients at zero initialization from finite nonzero learned-direction gradients, and does not require unsupported half-precision loss backward kernels.
- [ ] The bounded Example 07 DoRA helper executes composition, update and full-state reload on its declared dtype/device lanes; LoRA-only A/B loading is not counted as full DoRA persistence.
- [ ] The issue-specific changes in `docs/concepts.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FIX-002](https://github.com/thekaveh/NNx/issues/201), [FIX-003](https://github.com/thekaveh/NNx/issues/202), [FIX-013](https://github.com/thekaveh/NNx/issues/223).

## Delivery order and dependencies

- **Execution order:** 26 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FIX-003](https://github.com/thekaveh/NNx/issues/202), [FIX-026](https://github.com/thekaveh/NNx/issues/261). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. Independent of FIX-003: reproduced after explicitly converting the entire wrapper to half.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** `DoRALinear.forward` normalized the composed weight `V` in `V`'s own dtype with `clamp_min(1e-8)`. FP16 cannot represent `1e-8` (it rounds to `0`), so an all-zero row — after pruning, explicit zero init or a zero row in a loaded base — divided `0/0` and the whole output became NaN although `base(x)` was finite. Reproduced with the entire wrapper converted to half after wrapping, independent of parameter placement (FIX-003).

**What changed**
- The row norm, clamp, division and magnitude rescale now run in FP32 for FP16/BF16 layers (FP64 keeps FP64, FP32 unchanged) and the effective weight is cast back to the layer dtype before `F.linear`, so output dtype/device and the next half layer are unaffected. Gradients still flow through A, B and magnitude; a zero row stays exactly zero (never an invented direction); DoRA's DropConnect-style dropout on the composed update is unchanged. No epsilon knob, no serialized precision policy.
- Example 07: bounded `dora_zero_row_composition(dtype)` (`tests/test_examples_smoke.py -k dora_zero_row_composition`, run in FP16 **and** FP32) — mixed zero/nonzero base rows, `apply_dora_to`, whole-net conversion after wrapping, finite forward through the following projection, one SGD step on a float32 loss reduction moving only adapter/magnitude, bit-exact frozen base, full `state_dict()` reload including `magnitude`. Incidental: the example's pretrain and fine-tune phases produced the same run ID and `main()` failed with `FileExistsError` on develop (verified by running develop's file); they now carry distinct `data_id`s.
- Docs: `concepts.md` §11 DoRA bullet (zero/tiny-row behavior, FP32 normalization for FP16/BF16, full-state persistence, no blanket mixed-precision-kernel promise), `DoRALinear` docstring / API reference, examples catalog, CHANGELOG.

**PRs:** #281 (develop, merged), #282 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1638 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed). New: `test_dora_half_zero_rows_are_finite[bias=False|True]`, `test_dora_small_rows_match_promoted_reference[fp16|bf16|fp32|fp64]` (float64 oracle, FP64 not truncated), `test_dora_nonzero_learned_update_gradients[fp16|bf16|fp32|fp64]` (output + x/A/B/magnitude gradients vs oracle), `test_dora_half_classifier_step_and_full_state_round_trip`.
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- docs chain (`build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / projection tests) → pass; `python examples/07_lora_finetuning.py` runs end to end.
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #281 and #282.

**Not verifiable locally:** CPU FP16/BF16/FP32/FP64 lanes ran; no CUDA/MPS device was exercised on this host, so those backend cells are unverified (the fix is arithmetic-only and not backend-specific). Live Pages/wiki publication is left to the `docs.yml` main-push job.
